### PR TITLE
smart wallet report rejections of durable promises

### LINKED ISSUE #8997: refactor smartWallet error handling on executeOffer
## What is the Problem Being Solved?

In #8488, SmartWallet was refactored to use promiseWatcher to ensure that promises survived various kinds of upgrade trauma.

That refactoring left some error handling in a catch clause in `executeOffer`, missing the fact that that error handling would be lost when the walletFactory itself is upgraded. 

## Description of the Design

Move the [contents of the catch clause](https://github.com/Agoric/agoric-sdk/blob/695c440c0f48a3591b15a43665682c5f1ebbad9d/packages/smart-wallet/src/smartWallet.js#L992) in `executOffer` to appropriate sections of [`offerWatcher.js`](https://github.com/Agoric/agoric-sdk/blob/master/packages/smart-wallet/src/offerWatcher.js). 

## Security Considerations

Ensure that wallets correctly report on exceptions.

## Scaling Considerations

Not much relevance.

## Test Plan

It may be difficult to provoke the error conditions (multiple upgrades, and exceptions at carefully calibrated times would be required).  We aren't going to try to write tests for these corner cases, and will instead review carefully.

## Upgrade Considerations

This is about error behavior in upgrade situations. With the refactor, we expect the error handling to be invoked correctly even if there are vat upgrades.

### PR DESCRIPTION
**Pull request by turadg**
closes: #8997

## Description

This moves the error handling into an onRejected callback of the durable promise watcher for the offer result. The other two watchers are for particular values from Zoe after that promise resolves. Those are synchronous and don't need their own error handling.

It doesn't move `tryReclaimingWithdrawnPayments` because that would require heavier refactoring and in our analysis its a very unlikely scenario. I've commented with why it's there. I also changed its return type to `Promise<void>`, instead of all the settlements. Since that value was never read, it's not changing behavior and I called it a refactor.

### Security Considerations

No change.

### Scaling Considerations

No change.

### Documentation Considerations

It would help if Zoe documented in its code which methods' promises settle together.

### Testing Considerations

Ideally there'd be a new test for this fix but there isn't.

### Upgrade Considerations

To be included in u14 upgrade of smartWallet.

### CODE REVIEW

File: packages/smart-wallet/src/smartWallet.js [hunk 1]
```diff
@@ -990,21 +991,23 @@ export const prepareSmartWallet = (baggage, shared) => {
...
               facets.helper.updateStatus({
                 error: err.toString(),
                 ...offerSpec,
               });
             }
 
+            // Backstop recovery, in case something very basic fails.
             if (offerSpec?.proposal?.give) {
               facets.payments
                 .tryReclaimingWithdrawnPayments(offerSpec.id)
```
Comment (mhofman) [inline packages/smart-wallet/src/smartWallet.js:1025]: So we assume this case only matters for immediate rejections from the seat?

I suppose that `tryReclaimingWithdrawnPayments` is not very idempotent, nor parallel safe.

File: packages/smart-wallet/src/offerWatcher.js [hunk 2]
```diff
@@ -186,6 +193,17 @@ export const prepareOfferWatcher = baggage => {
               facets.helper.updateStatus({ result: UNPUBLISHED_RESULT });
           }
         },
+        /** @param {Error} err */
+        handleError(err) {
```
Comment (mhofman) [inline packages/smart-wallet/src/offerWatcher.js:202]: Maybe add a comment this will likely be called multiple times (aka it needs to handle parallels duplicate calls)

**Comment by mhofman** [comment]
> there may be unique error messages.

To be honest, while it's technically possible for a vstorage client to recover each of these errors, it's painful as the same field is updated, likely in the same block.

**Review by Chris-Hibbert — APPROVED** [review]
I don't think there's any value in reporting the errors 4 times. The different promises are all triggered on the same Zoe promise. If they fail independently, something is deeply wrong and having multiple watchers won't help.

I would drop the watcher in the original `catch` clause, as well as two of the three in offerWatcher. My nominee to keep would be `resultWatcher`, mostly because it sounds like it's at a high semantic level.

File: packages/smart-wallet/src/offerWatcher.js (hunk 2 above)
Comment (erights) [reply packages/smart-wallet/src/offerWatcher.js:202]: I don't get "parallel". Did you mean just "handle duplicate calls?"

File: packages/smart-wallet/src/offerWatcher.js [hunk 3]
```diff
@@ -209,6 +227,8 @@ export const prepareOfferWatcher = baggage => {
           const { facets } = this;
           if (isUpgradeDisconnection(err)) {
```
Comment (erights) [inline packages/smart-wallet/src/offerWatcher.js]: This line makes clear that the type of `err` must include `UpgradeDisconnection`. So probably `Error | UpgradeDisconnection`. Likewise elsewhere. The guards for all of these are fine though, since they say only `M.any()`.

File: packages/smart-wallet/src/smartWallet.js (hunk 1 above)
Comment (erights) [reply packages/smart-wallet/src/smartWallet.js:1025]: "parallel"?

Noting that at https://github.com/Agoric/agoric-sdk/pull/8488 I see 

> along the way `payments.tryReclaimingWithdrawnPayments()` went away. I'm pretty sure it was redundant with ERTP recoverySets.

File: packages/smart-wallet/src/offerWatcher.js (hunk 2 above)
Comment (mhofman) [reply packages/smart-wallet/src/offerWatcher.js:202]: There will be duplicate call, and since the async operation this initiates is not awaited, they will overlap

**Review by erights — APPROVED** [review]
@Chris-Hibbert and I walked through this verbally. LGTM!

We talked over the redundant errors being logged to vstorage. I think this is fine for now. The annoyance is minor, and I don't want to make further code changes at the last minute that we do not need.

File: packages/smart-wallet/src/smartWallet.js [hunk 4]
```diff
@@ -990,21 +991,23 @@ export const prepareSmartWallet = (baggage, shared) => {
             // await so that any errors are caught and handled below
             await watchOfferOutcomes(watcher, seatRef);
           } catch (err) {
-            facets.helper.logWalletError('OFFER ERROR:', err);
+            // This block only runs if the block above fails during one vat incarnation.
+            facets.helper.logWalletError('IMMEDIATE OFFER ERROR:', err);
 
-            // Notify the user
+            // Update status to observers
             if (err.upgradeMessage === 'vat upgraded') {
```
Comment (erights) [inline packages/smart-wallet/src/smartWallet.js:1010]: @Chris-Hibbert I just checked again, and this is the only non-test check manually using `upgradeMessage` rather than the `isUpgradeDisconnection`. I'm satisfied enough at our analysis that this is harmless to request it in this PR.

**Review by erights — CHANGES_REQUESTED** [review]

**Review by erights — APPROVED** [review]
I did a "request changes" as an interlock, to head off any concurrent attempt to merge. But approving now in case I'm hard to reach. In any case, the one remaining fix I'd like to see isn't crucial anyway. But would be nice.

**Review by gibson042 — APPROVED** [review]

File: packages/smart-wallet/src/offerWatcher.js [hunk 5]
```diff
@@ -220,15 +225,14 @@ export const prepareOfferWatcher = baggage => {
...
+         * If promise disconnected, watch again. Errors are handled by the resultWatcher.
          * @param {Error | import('@agoric/internal/src/upgrade-api.js').UpgradeDisconnection} reason
          * @param {UserSeat} seat
          */
         onRejected(reason, seat) {
           const { facets } = this;
           if (isUpgradeDisconnection(reason)) {
             void watchForPayout(facets, seat);
-          } else {
-            facets.helper.handleError(reason);
```
Comment (Chris-Hibbert) [inline packages/smart-wallet/src/offerWatcher.js]: Please revert this change. I mis-spoke when I said all the calls to `handleError` would get called at the same time. `resultWatcher` gets called separately from `numWantsWatcher` and `paymentWatcher`.  So in addition to calling `handleError` from `resultWatcher`, at least one of the others must be called, and this one reads better semantically.

File: packages/smart-wallet/src/offerWatcher.js [hunk 6]
```diff
@@ -186,6 +193,22 @@ export const prepareOfferWatcher = baggage => {
               facets.helper.updateStatus({ result: UNPUBLISHED_RESULT });
           }
         },
+        /**
+         * Called when the offer result promise rejects. The other two watchers
+         * are waiting for particular values out of Zoe but they settle at the same time
+         * and don't need their own error handling.
```
Comment (gibson042) [inline packages/smart-wallet/src/offerWatcher.js:199]:
```suggestion
         * Called when the offer result promise rejects. The payment and
         * numWantsSatisfied promises settle at the same time and don't need
         * their own error handling.
```

File: packages/smart-wallet/src/smartWallet.js (hunk 4 above)
Comment (Chris-Hibbert) [reply packages/smart-wallet/src/smartWallet.js:1010]: @turadg, since you're still making changes, would you mind updating this?
```suggestion
            if (isUpgradeDisconnection(err)) {
```

**Review by Chris-Hibbert — CHANGES_REQUESTED** [review]
Sorry, I requested a change while approving, and then everyone else approved. I don't want my request to get lost, so I'm changing my review status again.

And in the meantime, I requested another (very small) change.

File: packages/smart-wallet/src/smartWallet.js (hunk 4 above)
Comment (turadg) [reply packages/smart-wallet/src/smartWallet.js:1010]: oh yeah, I meant to but I don't think it's worth spinning CI at this point

**Review by Chris-Hibbert — APPROVED** [review]
Thanks. LGTM.

**Comment by mhofman** [comment]
Did this merge without new commits being pushed?
